### ISSUE TITLE
Error output on invalid tokens/options

### ISSUE THREAD

**Issue by targodan** [StabbyCutyou/moldova#14]
Heya, I just started playing with it and noticed some UX problem:

If you use an invalid token e.g. `'{doesNotExist}'` or if you have a tipo like `{firsname}` it just results in an empty string. This can be confusing, especially if you have a tiny tipo you don't immediately see.

Even worse, if you chose an invalid option, e.g. an unsupported language for a firstname it just seem to silently ignore that option, which results in the user wondering if this language is not supported or the names that are associated with that language are just not properly curated.

I would expect some error output like `Invalid token 'firsname' in column 12.` and `Invalid value 'obscure language here' for option 'language' of token 'firstname' in column 12.`.
The column output is not strictly necessary but would still be useful information, especially if you have lots of tokens in your template.

Ideally the output we have so far would still be output to stdout, but the error output would be output via the stderr not the stdout, so you can just pipe it to /dev/null if you are not interested in any error messages.

**Comment by StabbyCutyou** [comment]
Hey, sorry for the delay, I just noticed this issue now (and for some reason, I never get emails from GH).

I think these changes make sense. Let me work on a quick PR and I'll tag you in it.

**Comment by StabbyCutyou** [comment]
So, failing on an unknown token/word is pretty straightforward, i'll open the PR in a minute (I cleaned up the error handling in general with a bit of a newer approach i've taken in other libraries, which will let callers determine the type of error to take further action if need be).

The unknown language one, with how i've currently structured it, is a bit more difficult - it was designed so you could provide a list of "desired" languages, but that it would always fall-back to the default language if there were no language-specific spellings found. So it's sort of designed around the idea of always giving you something in the default language as a last resort, instead of failing.

I'll open the PR and you can provide feedback on the language issue there if you'd like

**Comment by targodan** [comment]
Heya, thx for reacting so quickly. 😃 

I can see why you wouldn't want to output errors on false languages. I would still argue that you should (Maybe prefix it as warning instead of error) because if you output the warning to stderr and the fallback output to stdout you could still easily pipe the stderr to /dev/null. Or if you pipe the output like `moldova ... | myslq ...` it will still work, as the stderr won't be redirected by default. However I see your point and won't hold it against you if you stick to it. 😉 

I will take a look at your PR and test it a little on the weekend.

**Comment by StabbyCutyou** [comment]
TL;DR; the logging to stderr is a workable idea, but to make it work smoothly for the library and the standalone binary, it will take a little refactoring.

The other concern here with just emitting to stderr is that Moldova is meant to function both as a library, and as a standalone binary.

So I don't want to just emit to stderr if something goes wrong from inside the library, ideally I'd only want to do that from the binary.

This means that I need to make it so people can pass in a logger to be used, so that the caller of the code can define "When an error happens but isn't stop-the-world bad, how do I learn about it?" via the logger.

So while I do think your idea works, it will take a bit of time to make it so the caller can define the log they want to emit to (because maybe they also want it to go to stdout for some reason, you never know).

I might punt on that part for now, in lieu of fixing the core issue you reported, but open a new PR once the current one is closed where i refactor and make it friendly to custom logging.

**Comment by StabbyCutyou** [comment]
Contradicting myself a bit, I've been thinking about it and I could see where, if you provided a language which we had 0 support for (ie, is not in the languages collection overall), that would be an error.

Under the current design, you've built the template wrong, but the system has no direct tangible way of informing you, other than to fail. But in this case, it doesn't. That seems wrong to me lol.

The logger abstraction would still work for other purposes, but this does feel like maybe I was wrong and it makes sense to return an error (when the language flat out doesn't exist in the system).

**Comment by StabbyCutyou** [comment]
Alright - I went with the hard error when a language doesn't exist at all, updated the PR - writing some tests now

**Comment by targodan** [comment]
Sorry for the delay, I had quite the week and a half.

Looks quite good to me. 👍 

To your point of using it both as a library and as a tool: I totally agree that a library should really not just output stuff to either stderr or stdout. However I think there is a pretty easy sollution (I haven't looked at your code yet). I would just have my render template function return an error but not output anything and then just have the error output to stderr inside of the cli-tool part of the code.

**Comment by StabbyCutyou** [comment]
That is how it works now, actually. My concern was to keep it that way in
the face of these changes, which i think I've accomplished.

If The code looks good, I can merge and do an official release tonight.

**Comment by targodan** [comment]
Brilliant! That should make your tool a lot more easily usable. 😃

**Comment by StabbyCutyou** [comment]
Released v2.0.0 since this sort of breaks the existing way the templates are handled. should be all set.
